debian/patches/0081-dovi-tonemapx.patch

```diff
@@ -0,0 +1,444 @@
+@@ -462,7 +669,8 @@ static void tonemap_frame_p016_p010_2_nv
+                                            int srcdepth,
+                                            int width,
+                                            int height,
+-                                           const struct TonemapIntParams *params)
++                                           const struct TonemapIntParams *params,
++                                           TonemapxContext *s)
+ {
+     const int in_depth = srcdepth;
+     const int in_uv_offset = 128 << (in_depth - 8);
+@@ -503,20 +711,58 @@ static void tonemap_frame_p016_p010_2_nv
+             int u = (srcuv[x]     >> in_sh2) - in_uv_offset;
+             int v = (srcuv[x + 1] >> in_sh2) - in_uv_offset;
+
+-            r[0] = av_clip_int16((y00 * cy + crv * v + in_rnd) >> in_sh);
+-            r[1] = av_clip_int16((y01 * cy + crv * v + in_rnd) >> in_sh);
+-            r[2] = av_clip_int16((y10 * cy + crv * v + in_rnd) >> in_sh);
+-            r[3] = av_clip_int16((y11 * cy + crv * v + in_rnd) >> in_sh);
+-
+-            g[0] = av_clip_int16((y00 * cy + cgu * u + cgv * v + in_rnd) >> in_sh);
+-            g[1] = av_clip_int16((y01 * cy + cgu * u + cgv * v + in_rnd) >> in_sh);
+-            g[2] = av_clip_int16((y10 * cy + cgu * u + cgv * v + in_rnd) >> in_sh);
+-            g[3] = av_clip_int16((y11 * cy + cgu * u + cgv * v + in_rnd) >> in_sh);
+-
+-            b[0] = av_clip_int16((y00 * cy + cbu * u + in_rnd) >> in_sh);
+-            b[1] = av_clip_int16((y01 * cy + cbu * u + in_rnd) >> in_sh);
+-            b[2] = av_clip_int16((y10 * cy + cbu * u + in_rnd) >> in_sh);
+-            b[3] = av_clip_int16((y11 * cy + cbu * u + in_rnd) >> in_sh);
++            float yuv1[3], yuv2[3], yuv3[3], yuv4[3], chroma_sample[3];
++            float c1[3], c2[3], c3[3], c4[3];
++
++            yuv1[0] = y00;
++            yuv2[0] = y01;
++            yuv3[0] = y10;
++            yuv4[0] = y11;
++            yuv1[1] = yuv2[1] = yuv3[1] = yuv4[1] = u;
++            yuv1[2] = yuv2[2] = yuv3[2] = yuv4[2] = v;
++
++            reshape_dovi_yuv(yuv1, yuv1, s);
++            reshape_dovi_yuv(yuv2, yuv2, s);
++            reshape_dovi_yuv(yuv3, yuv3, s);
++            reshape_dovi_yuv(yuv4, yuv4, s);
++            ycc2rgb(c1, yuv1[0], yuv1[1], yuv1[2], s->dovi->nonlinear, s->dovi->nonlinear_offset);
++            ycc2rgb(c2, yuv2[0], yuv2[1], yuv2[2], s->dovi->nonlinear, s->dovi->nonlinear_offset);
++            ycc2rgb(c3, yuv3[0], yuv3[1], yuv3[2], s->dovi->nonlinear, s->dovi->nonlinear_offset);
++            ycc2rgb(c4, yuv4[0], yuv4[1], yuv4[2], s->dovi->nonlinear, s->dovi->nonlinear_offset);
++            lms2rgb(c1, c1[0], c1[1], c1[2], s->dovi->linear);
++            lms2rgb(c2, c2[0], c2[1], c2[2], s->dovi->linear);
++            lms2rgb(c3, c3[0], c3[1], c3[2], s->dovi->linear);
++            lms2rgb(c4, c4[0], c4[1], c4[2], s->dovi->linear);
++
++            r[0] = av_clip_int16(c1[0] * 28672);
```

Comment:
This will make the picture look slightly correct, but it won't solve the problem.
`r[0] = av_clip_int16((int)(c1[0] * 28672) >> in_sh);`

In theory, this will give you the normalized pixel value for p010 input.
`pix_val / ((1<<10)-1) => 0.0f~1.0f`

But anyway the result of reshaped rgb output is wrong (I commented out tonemap_int16() to check)

This is really weird. I'll check it out again when I have some free time.

Reply:
I have investigated this a little bit more and found that:

For yuv420p10 inputs, we need to divide by 1023, to get the normalization, but wen we convert it back we need to multiple by 65535 to reinterpret it as a full 16bit color. the yuv420p10 inputs does not need to be shifted to low 10bits.

Current reshaping implementation seems to only reshape the intensity correctly and the chromas are not correctly reshaped. Doing `CLAMP(s, comp->pivots[0], comp->pivots[comp->num_pivots-1]);` on chromas will make all chromas to be a very small number (and mostly zero), the resulting video is mostly black-white. Not doing this the output video will behave like having a blue mask on it.